### ISSUE TITLE
cinnamon: HiDPI issues in menu and OSD

### ISSUE THREAD

**Issue by LinuxOnTheDesktop** [linuxmint/cinnamon#9015]
1) MintMenu (the menu applet): after force-enabling HiDPI in Settings, the menu sizes its second column poorly; see screenshot. 

![Screenshot from 2019-12-05 21-05-32](https://user-images.githubusercontent.com/47056543/70275215-8f555b80-17a5-11ea-8554-ddf248bf136d.png)

However, that sizing problem disappeared after a minute or so, in that when I next opened the menu, the problem was gone. (And the empty context-menu bug seems to be gone. That problem is reported [here](https://github.com/linuxmint/cinnamon/issues/7427).)

2) Having enable Hi-DPI (but not rebooted), the OCD icons - volume, brightness - were wrong (in that they were too large, and exceeded their background box, and seemed distorted). Also, *one* time that I switched from (I think it was) Hi-DPI to normal, icons on the desktop overlapped until I pressed F5. But this did not reoccur.

ThinkPad X1CG6. Live Cinnamon 19.3 session.
**Locale**:
```
mint@mint:~$ locale
LANG=C.UTF-8
LANGUAGE=
LC_CTYPE="C.UTF-8"
LC_NUMERIC="C.UTF-8"
LC_TIME="C.UTF-8"
LC_COLLATE="C.UTF-8"
LC_MONETARY="C.UTF-8"
LC_MESSAGES="C.UTF-8"
LC_PAPER="C.UTF-8"
LC_NAME="C.UTF-8"
LC_ADDRESS="C.UTF-8"
LC_TELEPHONE="C.UTF-8"
LC_MEASUREMENT="C.UTF-8"
LC_IDENTIFICATION="C.UTF-8"
LC_ALL=
```

**Comment by clefebvre** [comment]
We probably need to think design here... I'm not really keen on detecting scale change in every single applet out there.. we could show a popup saying the scale is about to change and then restart cinnamon to ensure everything works well.

Related somehow... we're getting feature requests for such a popup when a new screen is detected, and there are quite a few bugs we don't know how to fix which are related to HDMI/DP hotplug events. Restarting Cinnamon in these cases could guarantee a much more stable experience.

**Comment by nof8** [comment]
I have another HiDPI/normal screen switching menu issue which I also described: [in Cinnamon thread on Mint forum](https://forums.linuxmint.com/viewtopic.php?f=208&t=338608)

I've just bought an external 3840x2160 display & set it up in Cinnamon like HiDPI double scale. My laptop has 1920x1080 resolution.
I had a usual 1920x1080 external monitor earlier and swithing between two monitors was smooth - just plug in the connector or take it out.

After switching from HiDPI screen to normal screen main menu size becomes very tall, exceeding upper screen border. I just can't use search bar or click upper icons in menu.

I've attached a screenshot to illustrate how it looks like after switching from 3840x2160 to 1920x1080

![cinnamon-hidpi-menu-bug](https://user-images.githubusercontent.com/65599616/103947199-00e5a300-5138-11eb-8094-e0da0b62e8dc.jpg)

**Comment by nof8** [comment]
Wow, Mint 20.1 is available on the mirror sites without announcement yet!
Just checked the same issue on fresh Cinnamon 4.8.5. It's still there, menu is 2x tall after the switching from 3840x2160 2x to 1920x1080, but it's slightly less tall on Mint 20.1 Cinnamon Live USB so search box almost fits even with 1.2 font scaling as I usually do.
I think menu should stay the same half-screen size tall. But it's not so annoying now.

**Comment by nof8** [comment]
I've just upgraded to Linux Mint 20.1. The issue stays the same and it's still annoying :(

System: Kernel: 5.4.0-60-generic x86_64 bits: 64 compiler: gcc v: 9.3.0 Desktop: Cinnamon 4.8.5
wm: muffin dm: LightDM Distro: Linux Mint 20.1 Ulyssa base: Ubuntu 20.04 focal
Graphics: Device-1: Intel Skylake GT2 [HD Graphics 520] vendor: Lenovo driver: i915 v: kernel
bus ID: 00:02.0 chip ID: 8086:1916
Display: x11 server: X.Org 1.20.9 driver: modesetting unloaded: fbdev,vesa
resolution: 3840x2160~60Hz
OpenGL: renderer: Mesa Intel HD Graphics 520 (SKL GT2) v: 4.6 Mesa 20.2.6
direct render: Yes

![switch_from_hidpi_menu_bug](https://user-images.githubusercontent.com/23435526/104129594-d3ce0600-536c-11eb-9a4b-12b0542c5be9.jpg)

**Comment by mtwebster** [comment]
There's a pending pull request to address these things:
https://github.com/linuxmint/cinnamon/pull/9821

**Comment by nof8** [comment]
Thanks for information!

**Comment by nof8** [comment]
@mtwebster, will this PR also fix issues with icons?
You can see on the screenshot that some icons became very large after switching from 3840x2160 to 1920x1080 (e.g. Skype, Dropbox, Seafile, etc.)

Meanwhile I found workaround. When I enter Cinnamon panel settings and manually change icon size to one pixel larger and one pixel smaller (effecively leaving them at the same size) - icons correct their size. But why is it needed? Couldn't Cinnamon do it automatically on connecting/disconnecting monitor?

**Comment by mtwebster** [comment]
Yes it should - also, restarting cinnamon should resolve this as well (alt-f2, r, enter), and won't affect your session - this is simpler than changing icon sizes.

**Comment by nof8** [comment]
After using "Alt-f2, r, enter" during half a year I see that it's rather frequent when it takes up to 30 seconds to restart, rather annoying.

**Comment by nof8** [comment]
There's even simpler hotkey to restart cinnamon: Ctrl+Alt+Esc
But it'd be better if it'd be done automatically upon new display connection

**Comment by benmccann** [comment]
I've been trying to use restarting cinnamon as a workaround, but it leads to a number of issues that I think might not make it something we'd want to do as a permanent solution. E.g. my open windows all become reordered, which is a bit annoying as I try to keep them organized in a certain manner and lose track of where they all are

**Comment by benmccann** [comment]
@mtwebster @clefebvre I'll buy one of you a 4k monitor if you'd regularly use it with Cinnamon :smile: Serious offer. I'd love to see the kinks in HiDPI ironed out

**Comment by starlightknight** [comment]
This issue is insanely frustrating when using a KVM switch as the problem resurfaces every time you toggle machines on the KVM. There's been an open PR proposal since January that has been left behind. Is there anything we can do to help move this forward? 

I bought 4k monitors in January ironically, and this issue has lead to me barely taking advantage of 4k resolution on Linux because of how annoying the problem is, which has in turn increasingly lead to me considering switching distros. I've used Mint for over a decade and don't want to leave, so I'd rather help be part of the solution if possible.

What can we do to help move this forward?

**Comment by nof8** [comment]
I've realized that Ctrl+Alt+ESC is always faster and doesn't hang as "Alt+F2,r" does sometimes. Seems like these are different methods.
But, as  benmccann commented, it reorders windows, which can annoy someone.
